fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -563,9 +565,26 @@ RecordCoreException newPartlyBuildException(boolean continuedBuild,
                     .toString());
         }
 
+        validateTimeLimit();
         return MoreAsyncUtil.delayedFuture(toWait, TimeUnit.MILLISECONDS).thenApply(vignore3 -> true);
     }
 
+    private void validateTimeLimit() {
+        final long timeLimitMilliseconds = common.config.getTimeLimitMilliseconds();
+        if (timeLimitMilliseconds <= 0) {
+            return;
+        }
+        final long now = System.currentTimeMillis();
+        if (startingTimeMillis + timeLimitMilliseconds >= now) {
+            return;
+        }
+        throw new TimeLimitException("Time Limit Exceeded",
+                LogMessageKeys.TIME_STARTED, startingTimeMillis,
+                LogMessageKeys.TIME_LIMIT, timeLimitMilliseconds,
+                LogMessageKeys.TIME_ENDED, now,
+                LogMessageKeys.TIME_UNIT, "milliseconds");
```

Comment:
In theory, it might make sense to put `MILLIS` in the name of `TIME_STARTED`, `TIME_ENDED`, and (especially) `TIME_LIMIT` instead of having a separate `TIME_UNIT` field. The logic for that is mainly that someone looking at the logs might not know that `TIME_UNIT` is associated with the other fields in the log (especially as it's not associated with every time-based field).

Reply:
Changed.